src/composable/multiplexer.ts

```diff
@@ -0,0 +1,512 @@
+          } catch (e) {
+            throw new Error(`Error uploading to decentralized storage ${location}: ${e}`)
+          }
+        default:
+          throw new Error('Unsupported location')
+      }
+    }
+
+    return await data()
+      .then((d) => {
+        try {
+          // validate that `d` is a valid `bytes` ready to be abi-encoded
+          utils.hexlify(utils.arrayify(d))
+
+          // if we get here, we have a valid `data` field for the `ProofStruct`
+          // This means that if there was an upload function, it was called and the upload was successful
+          // note: we don't check if the location has changed because we don't care
+          this.location = location
+
+          return {
+            location,
+            data: d,
+          }
+        } catch (e) {
+          throw new Error(`data returned by uploader is invalid`)
+        }
+      })
+      .catch((e) => {
+        throw new Error(`Error preparing proof struct: ${e}`)
+      })
+  }
+
+  /**
+   * Poll a conditional order to see if it is tradeable.
+   * @param owner The owner of the conditional order.
+   * @param p The proof and parameters.
+   * @param chain Which chain to use for the ComposableCoW contract.
+   * @param provider An RPC provider for the chain.
+   * @param offChainInputFn A function, if provided, that will return the off-chain input for the conditional order.
+   * @throws If the conditional order is not tradeable.
+   * @returns The tradeable `GPv2Order.Data` struct and the `signature` for the conditional order.
+   */
+  static async poll(
+    owner: string,
+    p: ProofWithParams,
+    chain: SupportedChainId,
+    provider: providers.Provider,
+    offChainInputFn?: (owner: string, params: ConditionalOrderParams) => Promise<string>
+  ): Promise<[GPv2Order.DataStructOutput, string]> {
+    const contract = ComposableCoW__factory.connect(COMPOSABLE_COW_CONTRACT_ADDRESS[chain], provider)
+    const offChainInput = offChainInputFn ? await offChainInputFn(owner, p.params) : '0x'
+    return await contract.getTradeableOrderWithSignature(owner, p.params, offChainInput, p.proof)
+  }
+
+  /**
+   * The primary entry point for dumping the proofs and parameters for the conditional orders.
+   *
+   * This is to be used by watchtowers / indexers to store the proofs and parameters for the
+   * conditional orders off-chain. The encoding returned by this method may **NOT** contain all
+   * proofs and parameters, depending on the `filter` provided, and therefore should not be used
+   * to rehydrate the multiplexer from a user's perspective.
+   * @param filter {@link getProofs}
+   * @returns A JSON-encoded string of the proofs and parameters for the conditional orders.
+   */
+  dumpProofs(filter?: (v: string[]) => boolean): string {
+    const root = this.getOrGenerateTree().root
+    const payload = this.encodeToJSON(filter)
+    if (root !== this.getOrGenerateTree().root) throw new Error('cannot modify tree with filter')
```

Comment:
~Does `encodeToJSON` mutates the generated tree/root?~
~Doesn't seem so unless I missed it.~
~In that case, this comparison will never be true, or will it?~

Ok, I see now that the filter can mutate the state - which feels super sketchy to even consider this - but I think you had a good reason for adding it.

Maybe you are looking for the tree to be [a read only object](https://stackoverflow.com/questions/43120022/i-want-totally-immutable-object-in-ts)?
Or maybe not, as you do want to be able to replace the tree...

Maybe have the tree be replaceable but the tree object itself immutable?

I might be over thinking this, but feels to me this is a problem addressable at language level which wouldn't need to ensure a filter is mutating the underlying obj.

Reply:
Removed this check completely as it seems idiomatic with JS development that developers should be cognisant of the problems when supplying a callback function that modifies the very object that their filter callback is designed to filter :sweat_smile: